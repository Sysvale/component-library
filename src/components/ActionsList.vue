<template>
	<span
		id="actionList"
	>
		<div
			class="d-flex actionStyle"
			:class="position === 'right' ? 'justify-content-end' : 'justify-content-start'"
		>
			<div
				v-if="position === 'left'"
			>
				<div
					v-if="!itsBeignShown && actions.length > numberOfExpandedActions"
					class="actionRightBorder action p-4"
					@click="expandList"
				>
					{{ collapsedActionName }}
				</div>
				<div
					v-if="actions.length > numberOfExpandedActions && itsBeignShown"
					class="actionRightBorder action p-4"
					@click="collapseList"
				>
					{{ expandedActionName }}
				</div>
			</div>
			<div
				v-for="(action, i) in actions"
				:key="action.title"
			>
				<transition name="slide-fade">
					<div
						v-if="i <= internalnumberOfExpandedActions - 1"
						class="action p-4"
						:class="{ actionLeftBorder: i > 0 }"
						@click="$emit('actionClicked', action)"
					>
						<slot name="action" :list="action" />
					</div>
				</transition>
			</div>
			<div
				v-if="position === 'right'"
			>
				<div
					v-if="!itsBeignShown && actions.length > numberOfExpandedActions"
					class="actionLeftBorder action p-4"
					@click="expandList"
				>
					{{ collapsedActionName }}
				</div>
				<div
					v-if="actions.length > numberOfExpandedActions && itsBeignShown"
					class="actionLeftBorder action p-4"
					@click="collapseList"
				>
					{{ expandedActionName }}
				</div>
			</div>
		</div>
	</span>
</template>

<script>
import * as icons from 'vue-feather-icons'
export default {
	props: {
		actions: {
			type: Array,
			default: () => [],
			description: 'The list of the actions to be displayed.',
			required: true,
		},
		numberOfExpandedActions: {
			type: Number,
			default: 2,
			description: 'The number of items that will be displayed when the component is redered.',
			required: false,
		},
		collapsedActionName: {
			type: String,
			default: 'More actions',
			description: `The name that will be displayed in the last item of the actions list when it's collapsed.`,
			required: false,
		},
		expandedActionName: {
			type: String,
			default: 'Less actions',
			description: `The name that will be displayed in the last item of the actions list when it's expanded.`,
			required: false,
		},
		position: {
			type: String,
			default: 'right',
			description: `Speficies if the component will be rendered right side or left side.`,
			required: false,
		}
	},
	data() {
		return {
			action: _.cloneDeep(this.actions),
			internalnumberOfExpandedActions: this.numberOfExpandedActions,
			itsBeignShown: false,
		}
	},

	methods: {
		expandList() {
			this.internalnumberOfExpandedActions = this.action.length;
			this.itsBeignShown = !this.itsBeignShown;
		},

		collapseList() {
			this.internalnumberOfExpandedActions = this.numberOfExpandedActions;
			this.itsBeignShown = !this.itsBeignShown;
		},
	},
}
</script>
<style>
#actionList .actionLeftBorder {
	border-left: 1px solid #CED4DA;
}

#actionList .actionRightBorder {
	border-right: 1px solid #CED4DA;
}

#actionList .action {
	padding: 4px;
	cursor: pointer;
	border-radius: 1px;
}

#actionList .actionStyle {
	color: #6A7580;
	font-weight: 600;
}

#actionList .action:hover {
	background-color: #F8FAFD;
}

#actionList .slide-fade-enter-active {
	transition: all .3s ease;
}

#actionList .slide-fade-leave-active {
	transition: all .1s ease;
}

#actionList .slide-fade-enter, .slide-fade-leave-to {
	transform: translateX(10px);
	opacity: 0;
}
</style>
