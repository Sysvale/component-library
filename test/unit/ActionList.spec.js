import { mount, shallowMount, createLocalVue } from '@vue/test-utils';
import BootstrapVue from 'bootstrap-vue';
import ActionsList from '../../src/components/ActionsList.vue';
import lodash from 'lodash';

const localVue = createLocalVue();
localVue.use(BootstrapVue);

let actions = [
	{title: 'Icon1', img: 'img1'},
	{title: 'Icon2', img: 'img2'},
	{title: 'Icon3', img: 'img3'},
	{title: 'Icon4', img: 'img4'},
];

test('Component is mounted properly', () => {
	window._ = lodash;
	const wrapper = mount(ActionsList, {
			localVue,
			propsData: {
					actions: actions,
			},
	});

	expect(wrapper).toMatchSnapshot();
});

describe("Action list functioning", () => {
	test('If when the number of expanded actions to be shown is 1 and there is an array of length 4, are rendered 1 action and 1 button to show more', () => {
		window._ = lodash;
		const wrapper = mount(ActionsList, {
			localVue,
			propsData: {
				actions: actions,
				numberOfExpandedActions: 1,
				position: 'right',
			},
		});
	
		expect(wrapper.findAll('.action').length).toBe(2);
		expect(wrapper.findAll('.actionLeftBorder').length).toBe(1);
	});
	
	test('If when the number of expanded actions to be shown is 4 and there is an array of length 4, are rendered 4 actions', () => {
		window._ = lodash;
		const wrapper = mount(ActionsList, {
			localVue,
			propsData: {
				actions: actions,
				numberOfExpandedActions: 4
			},
		});
	
		expect(wrapper.findAll('.action').length).toBe(4);
	});
	
	test('If the list is expanded when the "more actions" button is clicked', async () => {
		window._ = lodash;
		const wrapper = mount(ActionsList, {
			localVue,
			propsData: {
				actions: actions,
				numberOfExpandedActions: 1,
				position: 'right',
			},
		});
	
	
		expect(wrapper.findAll('.action').length).toBe(2);

		wrapper
			.find('.actionLeftBorder')
			.trigger('click');
	
		await localVue.nextTick();
	
		expect(wrapper.findAll('.action').length).toBe(5);
	});

	test('If a event is emited when the action is clicked', () => {
		window._ = lodash;
		const wrapper = mount(ActionsList, {
			localVue,
			propsData: {
				actions: actions,
				numberOfExpandedActions: 4
			},
		});	
		wrapper.find('.action').trigger('click');

		expect(wrapper.emitted().actionClicked).toBeTruthy();
		expect(wrapper.emitted().actionClicked).toEqual([
			[
				{'img': 'img1', "title": 'Icon1'},
			]
		]);
	});
});
