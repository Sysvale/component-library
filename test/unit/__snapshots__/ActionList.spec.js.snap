// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component is mounted properly 1`] = `
<span id="actionList"><div class="d-flex actionStyle justify-content-end"><!----> <div><transition-stub name="slide-fade"><div class="action p-4"></div></transition-stub></div><div><transition-stub name="slide-fade"><div class="action p-4 actionLeftBorder"></div></transition-stub></div><div><transition-stub name="slide-fade"><!----></transition-stub></div><div><transition-stub name="slide-fade"><!----></transition-stub></div> <div><div class="actionLeftBorder action p-4">
				More actions
			</div> <!----></div></div></span>
`;
